File: scripts/sources/claude.py
"""ClaudeSource: discover and parse Claude Code session files.

Facts verified on real data (2026-08-13):
- Sessions: ~/.claude/projects/<slugified-cwd>/<uuid>.jsonl, one level deep
  (session subdirs like subagents/ are skipped). Honor CLAUDE_CONFIG_DIR.
- Titles: "ai-title" records (last occurrence wins).
- cwd / model: per-record fields on user/assistant records (model may switch
  mid-session; we collect the unique set).
- An assistant message is streamed across several records sharing
  message.id (thinking / text / tool_use); tool results arrive as user
  records whose content is only tool_result blocks — they must NOT open a
  new turn.
- Compaction: a system record with subtype "compact_boundary"; the whole
  pre-compact history stays in the file, so parsing resets at the last
  boundary and the follow-up summary message (isCompactSummary) is surfaced
  as a warning, not a user turn.
- Tool outputs are sometimes stubbed (<persisted-output>, content-replacement)
  — detect and warn instead of pretending the content is present.
"""
import json
import os
import re
from datetime import datetime
from typing import Dict, List, Optional, Set, Tuple

from . import Event, Session, SessionMeta, Source

UUID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")

# Records that never carry conversation content (silently skipped).
NOISE_TYPES = frozenset((
    "last-prompt", "mode", "permission-mode", "file-history-snapshot",
    "file-history-delta", "queue-operation",
))

# Meta/transcript-only flags: true means "not real conversation content".
SKIP_FLAGS = ("isMeta", "isCompactSummary", "isVirtual", "isVisibleInTranscriptOnly")

# User messages that are local-command echoes (/compact and friends).
LOCAL_CMD_PREFIXES = ("<command-name>", "<local-command-")


def default_projects_dir() -> str:
    base = os.environ.get("CLAUDE_CONFIG_DIR") or os.path.join(os.path.expanduser("~"), ".claude")
    return os.path.join(base, "projects")


def _iter_records(path: str) -> Tuple[List[dict], int]:
    """All JSON records + count of bad lines."""
    records, bad = [], 0
    with open(path) as fh:
        for line in fh:
            try:
                records.append(json.loads(line))
            except ValueError:
                bad += 1
    return records, bad


class ClaudeSource(Source):
    name = "claude"

    def __init__(self, projects_dir: Optional[str] = None) -> None:
        self.projects_dir = projects_dir or default_projects_dir()

    # --- discovery ---

    def _session_files(self) -> Dict[str, str]:
        """uuid -> path, one level under each project dir."""
        found: Dict[str, str] = {}
        if not os.path.isdir(self.projects_dir):
            return found
        for entry in os.listdir(self.projects_dir):
            sub = os.path.join(self.projects_dir, entry)
            if not os.path.isdir(sub):
                continue
            for name in os.listdir(sub):
                if not name.endswith(".jsonl"):
                    continue
                sid = name[:-6]
                if UUID_RE.match(sid):
                    found[sid] = os.path.join(sub, name)
        return found

    def _scan_meta(self, path: str) -> Tuple[Optional[str], Optional[str], Optional[str], Set[str]]:
        """title (ai-title last wins), cwd, started_at, models — no event parse."""
        title = cwd = started_at = None
        models: Set[str] = set()
        records, _ = _iter_records(path)
        for r in records:
            if r.get("type") == "ai-title" and r.get("aiTitle"):
                title = r["aiTitle"]
            elif r.get("type") in ("user", "assistant"):
                if cwd is None and r.get("cwd"):
                    cwd = r["cwd"]
                if started_at is None and r.get("timestamp"):
                    started_at = r["timestamp"]
                model = (r.get("message") or {}).get("model")
                if model:
                    models.add(model)
        return title, cwd, started_at, models

    def list_sessions(self, limit: int = 20) -> List[SessionMeta]:
        found = self._session_files()
        if not found:
            return []
        order = sorted(found.items(), key=lambda kv: os.path.getmtime(kv[1]), reverse=True)
        metas = []
        for sid, path in order[:limit]:
            title, cwd, started_at, models = self._scan_meta(path)
            metas.append(SessionMeta(
                id=sid, title=title, cwd=cwd, started_at=started_at,
                updated_at=datetime.fromtimestamp(os.path.getmtime(path)).isoformat(timespec="seconds"),
                model=", ".join(sorted(models)) or None,
            ))
        return metas

    # --- read ---

    def _find_file(self, session_id: str) -> Optional[str]:
        return self._session_files().get(session_id)

    def read_session(self, session_id: str) -> Session:
        path = self._find_file(session_id)
        if not path:
            raise LookupError(
                "未找到 Claude Code 会话 %s（已扫描 %s 下所有项目目录）。\n"
                "会话 ID 是 ~/.claude/projects/<项目>/*.jsonl 的文件名。" % (session_id, self.projects_dir))
        return self._parse(path, session_id)

    def _parse(self, path: str, session_id: str) -> Session:
        meta = SessionMeta(id=session_id)
        events: List[Event] = []
        calls: Dict[str, int] = {}   # tool_use_id -> index in events
        compacted = False
        warnings: List[str] = []
        models: Set[str] = set()
        unknown: Dict[str, int] = {}
        skipped_attachments = 0
        skipped_flags = 0
        skipped_local_cmd = 0

        records, bad = _iter_records(path)
        for r in records:
            t = r.get("type")
            if any(r.get(f) for f in SKIP_FLAGS):
                skipped_flags += 1
                continue
            if t == "user":
                if meta.cwd is None and r.get("cwd"):
                    meta.cwd = r["cwd"]
                if meta.started_at is None and r.get("timestamp"):
                    meta.started_at = r["timestamp"]
                skipped_local_cmd += self._handle_user(r, events, calls, warnings)
            elif t == "assistant":
                if meta.cwd is None and r.get("cwd"):
                    meta.cwd = r["cwd"]
                if meta.started_at is None and r.get("timestamp"):
                    meta.started_at = r["timestamp"]
                self._handle_assistant(r, events, calls, models)
            elif t == "ai-title":
                if r.get("aiTitle"):
                    meta.title = r["aiTitle"]
            elif t == "system":
                if r.get("subtype") == "compact_boundary":
                    events, calls = [], {}
                    compacted = True
                    warnings.append("会话在 %s 压缩过：已重置到压缩边界，仅保留摘要之后的上下文"
                                    % ((r.get("timestamp") or "?")[:16]))
            elif t == "attachment":
                skipped_attachments += 1
            elif t in NOISE_TYPES:
                pass
            else:
                unknown[t] = unknown.get(t, 0) + 1

        if bad:
            warnings.append("解析中跳过 %d 个坏行（并发写入或中断所致）" % bad)
        if skipped_flags:
            warnings.append("跳过 %d 条元数据/摘要记录（isMeta/isCompactSummary 等）" % skipped_flags)
        if skipped_local_cmd:
            warnings.append("跳过 %d 条本地命令回显（/compact 等）" % skipped_local_cmd)
        if skipped_attachments:
            warnings.append("跳过 %d 条 attachment 记录（hook、token 用量等）" % skipped_attachments)
        for t, n in sorted(unknown.items()):
            warnings.append("跳过未知记录类型 %s（%d 条）" % (t, n))

        if models:
            meta.model = ", ".join(sorted(models))
        return Session(meta=meta, events=events, compacted=compacted, warnings=warnings)

    # --- record handlers ---

    @staticmethod
    def _handle_user(r, events, calls, warnings) -> int:
        """Returns the number of local-command echoes skipped."""
        content = (r.get("message") or {}).get("content")
        if isinstance(content, str):
            text = content.strip()
            if not text:
                return 0
            if text.startswith(LOCAL_CMD_PREFIXES):
                return 1
            events.append(Event(kind="user_msg", role="user", text=text))
            return 0
        if not isinstance(content, list):
            return 0
        texts, results = [], []
        for b in content:
            if not isinstance(b, dict):
                continue
            bt = b.get("type")
            if bt == "text" and isinstance(b.get("text"), str):
                texts.append(b["text"])
            elif bt == "tool_result":
                results.append(b)
        if texts:
            events.append(Event(kind="user_msg", role="user",
                                text="\n".join(texts).strip()))
        for b in results:
            ClaudeSource._attach_tool_result(b, events, calls, warnings)
        return 0

    @staticmethod
    def _attach_tool_result(b, events, calls, warnings) -> None:
        tid = b.get("tool_use_id")
        content = b.get("content")
        if isinstance(content, list):
            parts = []
            for c in content:
                if isinstance(c, dict) and isinstance(c.get("text"), str):
                    parts.append(c["text"])
            text = "\n".join(parts)
        elif isinstance(content, str):
            text = content
        else:
            text = ""
        if text.startswith("<persisted-output>"):
            text = "<persisted-output>…（完整输出在 ~/.claude 的 tool-results 目录）"
            warnings.append("存在持久化输出 stub，工具输出未内联")
        if tid and tid in calls:
            events[calls[tid]].tool_output = text
        else:
            events.append(Event(kind="tool_output", text=text))
            warnings.append("存在无法配对 tool_use_id 的工具输出（已顺序追加）")

    @staticmethod
    def _handle_assistant(r, events, calls, models) -> None:
        msg = r.get("message") or {}
        model = msg.get("model")
        if model:
            models.add(model)
        content = msg.get("content")
        if not isinstance(content, list):
            return
        for b in content:
            if not isinstance(b, dict):
                continue
            bt = b.get("type")
            if bt == "thinking":
                thinking = b.get("thinking")
                if isinstance(thinking, str) and thinking.strip():
                    events.append(Event(kind="reasoning", role="assistant", text=thinking.strip()))
            elif bt == "text" and isinstance(b.get("text"), str) and b["text"].strip():
                events.append(Event(kind="assistant_msg", role="assistant", text=b["text"].strip()))
            elif bt == "tool_use":
                args = b.get("input")
                events.append(Event(kind="tool_call", text=b.get("name") or "?",
                                    tool_args=json.dumps(args, ensure_ascii=False) if args is not None else ""))
                if b.get("id"):
                    calls[b["id"]] = len(events) - 1
            # tool_use_2 = internal subagent calls; skipped on purpose

File: internal/sources/claude.go
package sources

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/HelloiOS2014/AgentRecovery/internal/core"
)

var claudeUUID = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

var noiseTypes = map[string]bool{
	"last-prompt": true, "mode": true, "permission-mode": true,
	"file-history-snapshot": true, "file-history-delta": true, "queue-operation": true,
}

var skipFlags = []string{"isMeta", "isCompactSummary", "isVirtual", "isVisibleInTranscriptOnly"}

func defaultClaudeProjects() string {
	if v := os.Getenv("CLAUDE_CONFIG_DIR"); v != "" {
		return filepath.Join(v, "projects")
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".claude", "projects")
}

type Claude struct {
	ProjectsDir string
}

func NewClaude(dir string) *Claude {
	if dir == "" {
		dir = defaultClaudeProjects()
	}
	return &Claude{ProjectsDir: dir}
}

func (c *Claude) Name() string                  { return "claude" }
func (c *Claude) LoadTitles() map[string]string { return map[string]string{} }

func (c *Claude) sessionFiles() (map[string]string, error) {
	found := map[string]string{}
	st, err := os.Stat(c.ProjectsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return found, nil
		}
		return nil, err
	}
	if !st.IsDir() {
		return found, nil
	}
	entries, err := os.ReadDir(c.ProjectsDir)
	if err != nil {
		return nil, err
	}
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		sub := filepath.Join(c.ProjectsDir, e.Name())
		files, err := os.ReadDir(sub)
		if err != nil {
			continue
		}
		for _, f := range files {
			name := f.Name()
			if f.IsDir() || !strings.HasSuffix(name, ".jsonl") {
				continue
			}
			sid := name[:len(name)-6]
			if claudeUUID.MatchString(sid) {
				found[sid] = filepath.Join(sub, name)
			}
		}
	}
	return found, nil
}

func (c *Claude) ListSessions(limit int) ([]core.SessionMeta, error) {
	files, err := c.sessionFiles()
	if err != nil {
		return nil, err
	}
	type pair struct {
		id, path string
		mtime    int64
	}
	var order []pair
	for sid, path := range files {
		info, err := os.Stat(path)
		if err != nil {
			continue
		}
		order = append(order, pair{sid, path, info.ModTime().UnixNano()})
	}
	sort.Slice(order, func(i, j int) bool { return order[i].mtime > order[j].mtime })
	if limit > 0 && len(order) > limit {
		order = order[:limit]
	}
	var metas []core.SessionMeta
	for _, p := range order {
		title, cwd, started, models := c.scanMeta(p.path)
		model := strings.Join(sortedKeys(models), ", ")
		info, _ := os.Stat(p.path)
		metas = append(metas, core.SessionMeta{
			ID: p.id, Title: title, Cwd: cwd, StartedAt: started,
			UpdatedAt: isoSec(info.ModTime()), Model: model,
		})
	}
	return metas, nil
}

func sortedKeys(m map[string]bool) []string {
	var k []string
	for s := range m {
		k = append(k, s)
	}
	sort.Strings(k)
	return k
}

func (c *Claude) scanMeta(path string) (title, cwd, started string, models map[string]bool) {
	models = map[string]bool{}
	rec, _, err := readJSONL(path)
	if err != nil {
		return
	}
	for _, r := range rec {
		switch asStr(r["type"]) {
		case "ai-title":
			if t := asStr(r["aiTitle"]); t != "" {
				title = t
			}
		case "user", "assistant":
			if cwd == "" {
				cwd = asStr(r["cwd"])
			}
			if started == "" {
				started = asStr(r["timestamp"])
			}
			if m := asStr(asMap(r["message"])["model"]); m != "" {
				models[m] = true
			}
		}
	}
	return
}

func (c *Claude) ReadSession(id string) (*core.Session, error) {
	files, err := c.sessionFiles()
	if err != nil {
		return nil, err
	}
	path := files[id]
	if path == "" {
		return nil, fmt.Errorf("%w: 未找到 Claude Code 会话 %s（已扫描 %s 下所有项目目录）", ErrNotFound, id, c.ProjectsDir)
	}
	return c.parse(path, id)
}

func skippedFlag(r map[string]any) bool {
	for _, f := range skipFlags {
		if asBool(r[f]) {
			return true
		}
	}
	return false
}

func (c *Claude) parse(path, id string) (*core.Session, error) {
	meta := core.SessionMeta{ID: id}
	var events []core.Event
	calls := map[string]int{}
	compacted := false
	var warnings []string
	models := map[string]bool{}
	unknown := map[string]int{}
	skippedAtt, skippedFlags, skippedLocal := 0, 0, 0

	rec, bad, err := readJSONL(path)
	if err != nil {
		return nil, err
	}
	for _, r := range rec {
		if skippedFlag(r) {
			skippedFlags++
			continue
		}
		t := asStr(r["type"])
		switch t {
		case "user":
			if meta.Cwd == "" {
				meta.Cwd = asStr(r["cwd"])
			}
			if meta.StartedAt == "" {
				meta.StartedAt = asStr(r["timestamp"])
			}
			skippedLocal += handleClaudeUser(r, &events, calls, &warnings)
		case "assistant":
			if meta.Cwd == "" {
				meta.Cwd = asStr(r["cwd"])
			}
			if meta.StartedAt == "" {
				meta.StartedAt = asStr(r["timestamp"])
			}
			handleClaudeAssistant(r, &events, calls, models)
		case "ai-title":
			if t := asStr(r["aiTitle"]); t != "" {
				meta.Title = t
			}
		case "system":
			if asStr(r["subtype"]) == "compact_boundary" {
				events, calls = nil, map[string]int{}
				compacted = true
				ts := asStr(r["timestamp"])
				if len(ts) > 16 {
					ts = ts[:16]
				}
				if ts == "" {
					ts = "?"
				}
				warnings = append(warnings, "会话在 "+ts+" 压缩过：已重置到压缩边界，仅保留摘要之后的上下文")
			}
		case "attachment":
			skippedAtt++
		default:
			if noiseTypes[t] {
				break
			}
			if t != "" {
				unknown[t]++
			}
		}
	}
	if bad > 0 {
		warnings = append(warnings, fmt.Sprintf("解析中跳过 %d 个坏行（并发写入或中断所致）", bad))
	}
	if skippedFlags > 0 {
		warnings = append(warnings, fmt.Sprintf("跳过 %d 条元数据/摘要记录（isMeta/isCompactSummary 等）", skippedFlags))
	}
	if skippedLocal > 0 {
		warnings = append(warnings, fmt.Sprintf("跳过 %d 条本地命令回显（/compact 等）", skippedLocal))
	}
	if skippedAtt > 0 {
		warnings = append(warnings, fmt.Sprintf("跳过 %d 条 attachment 记录（hook、token 用量等）", skippedAtt))
	}
	for t, n := range unknown {
		warnings = append(warnings, fmt.Sprintf("跳过未知记录类型 %s（%d 条）", t, n))
	}
	if len(models) > 0 {
		meta.Model = strings.Join(sortedKeys(models), ", ")
	}
	return &core.Session{Meta: meta, Events: events, Compacted: compacted, Warnings: warnings}, nil
}

func handleClaudeUser(r map[string]any, events *[]core.Event, calls map[string]int, warnings *[]string) int {
	content := asMap(r["message"])["content"]
	if s, ok := content.(string); ok {
		text := strings.TrimSpace(s)
		if text == "" {
			return 0
		}
		if strings.HasPrefix(text, "<command-name>") || strings.HasPrefix(text, "<local-command-") {
			return 1
		}
		*events = append(*events, core.Event{Kind: "user_msg", Role: "user", Text: text})
		return 0
	}
	list := asList(content)
	var texts []string
	var results []map[string]any
	for _, b := range list {
		m := asMap(b)
		switch asStr(m["type"]) {
		case "text":
			if t := asStr(m["text"]); t != "" {
				texts = append(texts, t)
			}
		case "tool_result":
			results = append(results, m)
		}
	}
	if len(texts) > 0 {
		*events = append(*events, core.Event{Kind: "user_msg", Role: "user", Text: strings.TrimSpace(strings.Join(texts, "\n"))})
	}
	for _, b := range results {
		attachToolResult(b, events, calls, warnings)
	}
	return 0
}

func attachToolResult(b map[string]any, events *[]core.Event, calls map[string]int, warnings *[]string) {
	tid := asStr(b["tool_use_id"])
	content := b["content"]
	var text string
	switch c := content.(type) {
	case string:
		text = c
	case []any:
		var parts []string
		for _, x := range c {
			if t := asStr(asMap(x)["text"]); t != "" {
				parts = append(parts, t)
			}
		}
		text = strings.Join(parts, "\n")
	}
	if strings.HasPrefix(text, "<persisted-output>") {
		text = "<persisted-output>…（完整输出在 ~/.claude 的 tool-results 目录）"
		*warnings = append(*warnings, "存在持久化输出 stub，工具输出未内联")
	}
	if tid != "" {
		if idx, ok := calls[tid]; ok {
			(*events)[idx].ToolOutput = text
			return
		}
	}
	*events = append(*events, core.Event{Kind: "tool_output", Text: text})
	*warnings = append(*warnings, "存在无法配对 tool_use_id 的工具输出（已顺序追加）")
}

func handleClaudeAssistant(r map[string]any, events *[]core.Event, calls map[string]int, models map[string]bool) {
	msg := asMap(r["message"])
	if m := asStr(msg["model"]); m != "" {
		models[m] = true
	}
	for _, b := range asList(msg["content"]) {
		m := asMap(b)
		switch asStr(m["type"]) {
		case "thinking":
			th := strings.TrimSpace(asStr(m["thinking"]))
			if th != "" {
				*events = append(*events, core.Event{Kind: "reasoning", Role: "assistant", Text: th})
			}
		case "text":
			t := strings.TrimSpace(asStr(m["text"]))
			if t != "" {
				*events = append(*events, core.Event{Kind: "assistant_msg", Role: "assistant", Text: t})
			}
		case "tool_use":
			*events = append(*events, core.Event{
				Kind: "tool_call", Text: orQ(asStr(m["name"])), ToolArgs: jsonDump(m["input"]),
			})
			if id := asStr(m["id"]); id != "" {
				calls[id] = len(*events) - 1
			}
		}
	}
}

func orQ(s string) string {
	if s == "" {
		return "?"
	}
	return s
}
